### ISSUE TITLE
CI/CD Example Please

### ISSUE THREAD

**Issue by suntong** [kubesail/qotm#2]
Hi, 

Thanks for providing a Free plan to allow us to start trying kubernete! 
I've created the simple nignix web service and qotm, following the web prompt, successfully. 

However, when it comes to Automated deployments to Kubernetes with CI/CD, I'm all at lost because everything is new to me. I did have found an article:

**Automated deployments to Kubernetes with GitLab**  
https://sanderknape.com/2019/02/automated-deployments-kubernetes-gitlab/

However I'm only able to follow it half way till docker automated builds, but not deployments to Kubernetes, because the article is not using kubesail. 

So I'm wondering if you could give us a detailed example, like above, so that even for newbies like me can follow. 

Oh, just found another way, 

**Adding an existing Kubernetes cluster to gitlab**  
https://gitlab.com/help/user/project/clusters/index#adding-an-existing-kubernetes-cluster

How do we setup that way? 

Thanks a lot!

**Comment by pastudan** [comment]
Hey @suntong, thanks for trying out KubeSail!

**TLDR**: Sorry for taking a while to respond, I wanted to try to get GitLab CI working on my end before I replied. Unfortunately I don't think it will work as-is without some changes on the GitLab end. I will reach out to them and see if they would be willing to make some changes to allow it to work on multi-tenant kubernetes clusters, like KubeSail.

---

The long answer:

For security reasons, we can't give you access to the `kubectl cluster-info` command, as it could be used to see other users deployments. **However**, the info you need from it is available at https://kubesail.com/config. There are some steps to extract that info to be usable in GitLab:

From your kubectl config, the GitLab **API URL** field comes from `clusters[0].cluster.server`, in my case its `https://c1.kubesail.com:6443`

**CA Certificate** comes from `users[0].user['client-key-data']` but needs to be base64 decoded. You can copy / paste it into something like https://www.base64decode.org/

**Service Token** comes from `users[0].user['token']` and can be copied directly

**project namespace** is your github username, `suntong`

With all that said, and I was able to get GitLab talking to KubeSail, but their CI errors at the `$ ensure_namespace` step since we don't allow you to list or edit namespaces. This shouldn't be necessary, so I'll reach out to GitLab and see if they can make that step optional. If not, we may be able to modify the response to only list your namespace. I'll follow up here.

**Comment by suntong** [comment]
Totally understand. Thanks for trying to get it working on your end. 

Would github be any better, if GitLab needs to be changes on their end? Do you have a .travis.yml file that works or at least half-working? I read that fully automated deployments are still not 100% available for kubesail at the moment. Hope it will be one day. 

Thanks for all you helps, as everything is new to me and I couldn't figure those things out myself, especially if it is even not possible.

**Comment by pastudan** [comment]
I tried to get it working by using the Auto DevOps pipeline. I suspect that if you copy that yaml template (linked below) you could remove the `ensure_namespace` steps and get it working. I'd love to do a full blog post on this soon, but it may be a while before I have the time to get to that.

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/lib/gitlab/ci/templates/Auto-DevOps.gitlab-ci.yml

and specifically

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/lib/gitlab/ci/templates/Jobs/Deploy.gitlab-ci.yml

**Comment by suntong** [comment]
thanks Dan, will check them out and get back to you...

**Comment by suntong** [comment]
Hmm... I [tried](https://gitlab.com/suntong/hello-web-k8s/commits/master) but don't even know where I failed. So I guess instead of making things work, I have to wait and see a working version then adapt... 

take your time but looking forward to your full blog...
